### PR TITLE
PML/UCX: improved error processing in MPI_Recv

### PR DESCRIPTION
**Pull request by hoopoepg**
- improved error processing in MPI_Recv implementation
  of pml UCX

Signed-off-by: Sergey Oblomov <sergeyo@mellanox.com>

### CODE REVIEW

**Review by brminich — APPROVED** [review]

File: ompi/mca/pml/ucx/pml_ucx.c [hunk 1]
```diff
@@ -627,15 +630,15 @@ int mca_pml_ucx_recv(void *buf, size_t count, ompi_datatype_t *datatype, int src
     MCA_COMMON_UCX_PROGRESS_LOOP(ompi_pml_ucx.ucp_worker) {
         status = ucp_request_test(req, &info);
         if (status != UCS_INPROGRESS) {
-            mca_pml_ucx_set_recv_status_safe(mpi_status, status, &info);
+            mca_pml_ucx_set_recv_status(mpi_status, status, &info);
```
Comment (yosefe) [inline ompi/mca/pml/ucx/pml_ucx.c:633]: maybe add return value for mca_pml_ucx_set_recv_status(), instead of extra branch in lines 615-616?

Comment (hoopoepg) [reply ompi/mca/pml/ucx/pml_ucx.c:633]: branch on 615-616 needed to call mca_pml_ucx_set_recv_status. I can move it in to mca_pml_ucx_set_recv_status_safe

**Comment by hoopoepg** [comment]
@yosefe ok to squash?

File: ompi/mca/pml/ucx/pml_ucx_request.h [hunk 2]
```diff
@@ -186,15 +186,19 @@ static inline void mca_pml_ucx_set_recv_status(ompi_status_public_t* mpi_status,
...
+                                                   const ucp_tag_recv_info_t *info)
 {
-    if (mpi_status != MPI_STATUS_IGNORE) {
-        mca_pml_ucx_set_recv_status(mpi_status, ucp_status, info);
-    }
+    ompi_status_public_t _local_status;
+    ompi_status_public_t *mpi_status = (_mpi_status != MPI_STATUS_IGNORE) ?
+                                       _mpi_status : &_local_status;
+
+    return mca_pml_ucx_set_recv_status(mpi_status, ucp_status, info);
```
Comment (yosefe) [inline ompi/mca/pml/ucx/pml_ucx_request.h:201]: maybe avoid setting temp mpi_status field if not needed
```
if (mpi_status != MPI_STATUS_IGNORE) {
    return mca_pml_ucx_set_recv_status(mpi_status, ucp_status, info)
} else if (status == UCS_OK || status == CANCELED) {
   return UCS_OK;
} else if (status == TRUNCATED)
  reutrn ERR_TRUNCATE
} else {
 return ERR_INTERN
}
```

File: ompi/mca/pml/ucx/pml_ucx.c [hunk 3]
```diff
@@ -627,15 +628,15 @@ int mca_pml_ucx_recv(void *buf, size_t count, ompi_datatype_t *datatype, int src
     MCA_COMMON_UCX_PROGRESS_LOOP(ompi_pml_ucx.ucp_worker) {
         status = ucp_request_test(req, &info);
         if (status != UCS_INPROGRESS) {
-            mca_pml_ucx_set_recv_status_safe(mpi_status, status, &info);
+            result = mca_pml_ucx_set_recv_status_safe(mpi_status, status, &info);
```
Comment (yosefe) [inline ompi/mca/pml/ucx/pml_ucx.c:631]: do we need this also in other places mca_pml_ucx_set_recv_status_safe is used?

File: ompi/mca/pml/ucx/pml_ucx_request.h (hunk 2 above)
Comment (hoopoepg) [reply ompi/mca/pml/ucx/pml_ucx_request.h]: updated

File: ompi/mca/pml/ucx/pml_ucx.c (hunk 3 above)
Comment (hoopoepg) [reply ompi/mca/pml/ucx/pml_ucx.c:631]: fixed for MPI_Mrecv call, other locations are non-blocking MPI calls - not relevant

**Comment by hoopoepg** [comment]
@yosefe ok to squash?

File: ompi/mca/pml/ucx/pml_ucx_request.h [hunk 4]
```diff
@@ -186,15 +186,23 @@ static inline void mca_pml_ucx_set_recv_status(ompi_status_public_t* mpi_status,
...
-                                                    ucs_status_t ucp_status,
-                                                    const ucp_tag_recv_info_t *info)
+static inline int mca_pml_ucx_set_recv_status_safe(ompi_status_public_t* mpi_status,
+                                                   ucs_status_t ucp_status,
+                                                   const ucp_tag_recv_info_t *info)
 {
     if (mpi_status != MPI_STATUS_IGNORE) {
-        mca_pml_ucx_set_recv_status(mpi_status, ucp_status, info);
+        return mca_pml_ucx_set_recv_status(mpi_status, ucp_status, info);
+    } else if ((ucp_status == UCS_OK) || (ucp_status == UCS_ERR_CANCELED)) {
```
Comment (yosefe) [inline ompi/mca/pml/ucx/pml_ucx_request.h:199]: maybe add opal_likely?

Comment (hoopoepg) [reply ompi/mca/pml/ucx/pml_ucx_request.h]: ok, added

**Comment by hoopoepg** [comment]
@yosefe ok to squash?

**Comment by hoopoepg** [comment]
@yosefe ok to merge?

**Review by yosefe — APPROVED** [review]

**Pull request merged**